functorch/functorch/experimental/cond.py

```diff
@@ -0,0 +1,158 @@
+from torch.utils._pytree import tree_flatten
+from torch.fx.experimental.proxy_tensor import get_isolated_graphmodule, disable_proxy_modes_tracing, get_proxy_slot
+import torch.utils._pytree as pytree
+from torch.dispatch.dispatcher import dispatcher_singleton
+from torch.utils._mode_utils import no_dispatch
+import random
+import string
+
+"""
+We're going to define a `cond` operation.
+In order to do this, we need implementations for each of the dispatch keys.
+"""
+from contextlib import contextmanager
+
+# TODO(voz): Move out somewhere else once other py dispatched ops need it
+@contextmanager
+def suspend_mode(mode):
+    assert(mode is not None), "Cannot suspend None mode"
+    torch._C._set_torch_dispatch_mode(None)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(mode)
+
+@contextmanager
+def enable_mode(mode):
+    curr_mode = torch._C._get_torch_dispatch_mode()
+    torch._C._set_torch_dispatch_mode(mode)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(curr_mode)
+
+
+def trace_cond(proxy_mode, func_overload, args, kwargs=None):
+    assert kwargs is None or not kwargs
+    pred, true_fn, false_fn, operands = args
+
+    def _unwrap_proxy(e):
+        proxy_or_e = get_proxy_slot(e, proxy_mode.tracer, e, lambda e: e.proxy )
+        return proxy_or_e
+
+
+    if isinstance(operands, torch.Tensor):
+        operands = [operands]  # Little hack because * on a single ProxyTensor unpacks it
+    else:
+        operands = operands
+
+    true_graph = get_isolated_graphmodule(true_fn, operands, {})
+    false_graph = get_isolated_graphmodule(false_fn, operands, {})
+
+    # There are probably better ways - I know that create_arg has some self incrementing name
+    # magic to it, but since we explicitly have to get the name for register_module,
+    # I was not sure how to do that. This kinda simulates it.
+    next_name = None
+    i = 0
+    while not next_name:
+        candidate = f"true_graph_{i}"
+        if hasattr(proxy_mode.tracer.root, candidate):
+            i += 1
+        else:
+            next_name = candidate
+
+    random_slug = ''.join(random.choices(string.digits, k=5))
+    true_name = next_name
+    false_name = f"false_graph_{i}"
+
+    proxy_mode.tracer.root.register_module(true_name, true_graph)
+    proxy_mode.tracer.root.register_module(false_name, false_graph)
+
+    with no_dispatch():
+        # This is not amazing.
+        # However, if we have nested operators that have a call_function
+        # in their graph that is not a torch op (ex: see conditional below, nested cond)
+        # we cannot get metadata for it from just looking at out vars.
+        # The reason is that an operation on the output of such an op is not
+        # evalauted as a torch.Tensor.
+        # So we execute the real true and false fn here and compare metadata
+        true_result = true_fn(*operands)
+        false_result = false_fn(*operands)
```

Comment:
I wish there were some way to do this at the same time as when we are tracing out the true_graph and the false_graph. When one does a make_fx trace of true_fn, then at some point, we do obtain the output of true_fn (because we are tracing using a Proxy and a concrete Tensor); ditto for false_fn; and then we wouldn't need to re-evaluate true_fn and false_fn to compare their outputs.

This doesn't matter as much when we trace in other modes (i.e., fake, symbolic), since the re-execution should be relatively quicker due to not having data

Reply:
Agreed - I had this as such, but the problem is that the output of the call_function I wrote is not recognized as a torch.tensor, so all downstream operations do not get converted to torch ops correctly in tracing.

Rough ex from memory:

```
def foo(y):
     x = cond(...)
     return y + x
```     
     
would get traced as 

call_function -> 'cond'
call_function -> 'add'

versus 

```
def foo(y):
     x = torch.randn(...)
     return y + x
```

would get traced as

call_function -> 'torch.ops.randn'
call_function -> 'torch.ops.add'

From the second example, it was possible to access the traced meta by walking through the output nodes in the traced graph. However, on the first, that 'add' is (1) Not an op class, but rather a string and (2) has no meta